tools/clean-venv-cache

```diff
@@ -1,28 +1,68 @@
 #!/usr/bin/env python
 from __future__ import absolute_import
 from __future__ import print_function
-import glob
+import argparse
+import datetime
 import os
 import subprocess
-import sys
-
-ZULIP_PATH = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
-VENV_CACHE_DIR = "/srv/zulip-venv-cache"
-if "--travis" in sys.argv:
-    VENV_CACHE_DIR = os.path.join(os.environ["HOME"], "zulip-venv-cache")
-
-used_caches = set()
-hash_reqs = os.path.join(ZULIP_PATH, 'scripts', 'lib', 'hash_reqs.py')
-
-for filename in os.listdir(os.path.join(ZULIP_PATH, "requirements")):
-    requirements_file = os.path.join(ZULIP_PATH, "requirements", filename)
-    hash_val = subprocess.check_output([hash_reqs, requirements_file]).strip()
-    used_caches.add(os.path.join(VENV_CACHE_DIR, hash_val.decode('utf-8')))
-
-for cache_dir_base in os.listdir(VENV_CACHE_DIR):
-    cache_dir = os.path.join(VENV_CACHE_DIR, cache_dir_base)
-    if cache_dir not in used_caches:
-        print("Cleaning unused venv %s" % (cache_dir,))
-        subprocess.check_call(["sudo", "rm", "-rf", cache_dir])
-    else:
-        print("Keeping used venv %s" % (cache_dir,))
+import time
+
+DEPLOYMENTS_DIR = "/home/zulip/deployments"
+
+def parse_args():
+    parser = argparse.ArgumentParser(description="This script cleans unused zulip venv caches.")
+    parser.add_argument("--env", dest="env", type=str, default="prod", nargs="?",
+                        choices=["travis", "dev", "prod"], metavar="<env>",
+                        help="The environment(travis, dev or prod) in which this script is being executed.(defaults to prod)")
+    parser.add_argument("--threshold", dest="threshold_days", type=int, default=14, nargs="?", metavar="<days>",
+                        help="Caches older than the specified number of days will be deleted.(defaults to 14)")
+    args = parser.parse_args()
+    return args
+
+def get_caches_to_purge(venv_dir, env, threshold_days):
+    to_purge = []
+    used_caches = set()
+
+    if env in ["dev", "travis"]:
+        ZULIP_PATH = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+        hash_reqs = os.path.join(ZULIP_PATH, 'scripts', 'lib', 'hash_reqs.py')
+        for filename in os.listdir(os.path.join(ZULIP_PATH, "requirements")):
+            requirements_file = os.path.join(ZULIP_PATH, "requirements", filename)
+            hash_val = subprocess.check_output([hash_reqs, requirements_file]).strip()
+            used_caches.add(os.path.join(venv_dir, hash_val.decode('utf-8')))
+
+    if env == "prod":
+        for deployment in ["current", "next", "last"]:
+            curr_venv_cache = os.path.join(DEPLOYMENTS_DIR, deployment, "zulip-current-venv")
+            used_caches.add(curr_venv_cache)
+
+    threshold = datetime.datetime.now() - datetime.timedelta(threshold_days)
+    threshold_timestamp = time.mktime(threshold.utctimetuple())
```

Comment:
For prod, I think we actually want to be checking the threshhold on the timestamp for the deployment, not the timestamp of the venv.  Because an older virtualenv could be used by 3 deployments in a row, the last of which is only 2 days old, and if we garbage-collect the venv, we'll break that 2-day-old deployment.  

In effect, what this means is that we should probably have a function, perhaps `get_recent_deployments(threshhold_days)` in `scripts/lib/zulip_tools.py`, that takes a threshhold and returns a list of venv directories, combining last/next/current (actually, a better version would probably be "any directory that isn't a timestamp", to automatically handle anything like last/next/current we add in the future), `root/zulip` (if it exists), and ones with a timestamp more recent than the threshhold.  In the prod case, this script (and its siblings) would just call that function and loop through them.

A final note is that if we do this, we can actually make the dev/prod code paths have a lot more code in common.  Basically:
* First we get the list of directories (using `get_recent_deployments`, in prod; and add the current directory unconditionally for dev/travis)
* Then we hash all the requirements files in all those directories to get `used_caches`
* Then we add `threshhold` to save any that are new.  This is valuable in case someone is doing something with a Zulip checkout not in the `deployments/` directory for testing or whatever.
* Then we purge.

I think that pattern would work well for the other caches as well.